[INLONG-7061][Sort] Support table level metrics for Apache Doris connector and add dirty metrics

Yizhou-Yang opened the pull request:
Fixes #7061

Design:
for all incoming dirty data, try to parse and add table level metrics, no matter what the data is. If it fails for any reason, add global metrics instead.

the try block only works if 
1) the dirty data is rowdata, if deserialization causes error/npe, then just add global metric
2) the task is multiple sink，since table level metric is meaningless for single table tasks.

for single table and unparsable dirty data, either null pointer or ioexception will be thrown. In either case, just add node level metrics instead.

yunqingmoswu commented on inlong-sort/sort-connectors/doris/src/main/java/org/apache/inlong/sort/doris/table/DorisDynamicSchemaOutputFormat.java:
```diff
@@ -486,6 +489,10 @@ private void handleDirtyData(Object dirtyData, DirtyType dirtyType, Exception e)
             throw ex;
         }
+
+        long size = dirtyData.toString().getBytes(StandardCharsets.UTF_8).length;
```
Dirty data metrics is needed based on table metrics.

EMsnap commented on inlong-sort/sort-connectors/doris/src/main/java/org/apache/inlong/sort/doris/table/DorisDynamicSchemaOutputFormat.java:
```diff
@@ -486,6 +489,10 @@ private void handleDirtyData(Object dirtyData, DirtyType dirtyType, Exception e)
         }
+
+        long size = dirtyData.toString().getBytes(StandardCharsets.UTF_8).length;
+        metricData.invokeDirty(1, size);
```
npe

gong commented:
you need add dirty metric like out metric
```java
  if (multipleSink) {
                        String[] tableWithDb = tableIdentifier.split("\\.");
                        metricData.outputMetricsWithEstimate(tableWithDb[0], null, tableWithDb[1],
                                false, respContent.getNumberLoadedRows(), respContent.getLoadBytes());
                    } else {
                        metricData.invoke(respContent.getNumberLoadedRows(), respContent.getLoadBytes());
                    }
```

Yizhou-Yang commented:
> you need add dirty metric like out metric
> 
> ```java
>   if (multipleSink) {
>                         String[] tableWithDb = tableIdentifier.split("\\.");
>                         metricData.outputMetricsWithEstimate(tableWithDb[0], null, tableWithDb[1],
>                                 false, respContent.getNumberLoadedRows(), respContent.getLoadBytes());
>                     } else {
>                         metricData.invoke(respContent.getNumberLoadedRows(), respContent.getLoadBytes());
>                     }
> ```

thx, added a fix for whole db similar to this one

yunqingmoswu commented on inlong-sort/sort-connectors/doris/src/main/java/org/apache/inlong/sort/doris/table/DorisDynamicSchemaOutputFormat.java:
```diff
@@ -486,6 +497,17 @@ private void handleDirtyData(Object dirtyData, DirtyType dirtyType, Exception e)
         }
+
+        if (multipleSink) {
+            String[] tableWithDb = tableIdentifier.split("\\.");
```
It is not right because of the tableIdentifer is dynamic when it it multiplesink.

Yizhou-Yang replied:
the table identifier.split is used for multiple sink regular out data count, so I think it works for dirty data too...? or is there something i missed

yunqingmoswu commented on inlong-sort/sort-connectors/doris/src/main/java/org/apache/inlong/sort/doris/table/DorisDynamicSchemaOutputFormat.java:
```diff
@@ -486,6 +497,17 @@ private void handleDirtyData(Object dirtyData, DirtyType dirtyType, Exception e)
             throw ex;
         }
+
+        if (multipleSink) {
```
It is suggest to send the metic after the dirty data sink because the dirty sink maybe failed.

gong commented on inlong-sort/sort-connectors/base/src/main/java/org/apache/inlong/sort/base/metric/sub/SinkTableMetricData.java:
```diff
@@ -50,6 +50,11 @@ public class SinkTableMetricData extends SinkMetricData implements SinkSubMetric
 
+    /**
+     * The sub sink metric data container of dirty sink metric data
+     */
+    private final Map<String, SinkTableMetricData> subSinkDirtyMap = Maps.newHashMap();
+
```
not need add it, just use `subSinkMetricMap`.

gong commented on inlong-sort/sort-connectors/doris/src/main/java/org/apache/inlong/sort/doris/table/DorisDynamicSchemaOutputFormat.java:
```diff
@@ -503,6 +511,20 @@ private void handleDirtyData(Object dirtyData, DirtyType dirtyType, Exception e)
+
+        if (multipleSink) {
+            if (DirtyType.DESERIALIZE_ERROR.equals(dirtyType) || !(dirtyData instanceof RowData)) {
+                metricData.outputDirtyMetricsWithEstimate(null, null, null, rowSize, dataSize);
```
why not to pass databse table ?

Yizhou-Yang replied:
bacause there is no database/table avaliable to parse since the dirty data can't be deserialized, so only add global count in this case

gong commented on inlong-sort/sort-connectors/doris/src/main/java/org/apache/inlong/sort/doris/table/DorisDynamicSchemaOutputFormat.java:
```diff
@@ -503,6 +511,20 @@ private void handleDirtyData(Object dirtyData, DirtyType dirtyType, Exception e)
+            if (DirtyType.DESERIALIZE_ERROR.equals(dirtyType) || !(dirtyData instanceof RowData)) {
+                metricData.outputDirtyMetricsWithEstimate(null, null, null, rowSize, dataSize);
+            }
+            multipleDirtyHandler(dirtyData, rowSize, dataSize);
```
duplication code with `metricData.outputDirtyMetricsWithEstimate(null, null, null, rowSize, dataSize);`

gong commented on inlong-sort/sort-connectors/base/src/main/java/org/apache/inlong/sort/base/metric/sub/SinkTableMetricData.java:
```diff
@@ -199,6 +199,25 @@ public void outputMetricsWithEstimate(String database, String schema, String tab
+            long rowCount, long rowSize) {
+        if (StringUtils.isBlank(database) || StringUtils.isBlank(table)) {
+            invokeDirty(rowCount, rowSize);
+            return;
```
lost `StringUtils.isBlank(schema)`

gong commented on inlong-sort/sort-connectors/doris/src/main/java/org/apache/inlong/sort/doris/table/DorisDynamicSchemaOutputFormat.java:
```diff
@@ -512,17 +512,12 @@ private void handleDirtyData(Object dirtyData, DirtyType dirtyType, Exception e)
+            JsonNode rootNode = jsonDynamicSchemaFormat.deserialize(((RowData) dirtyData).getBinary(0));
+            metricData.outputDirtyMetricsWithEstimate(
+                    jsonDynamicSchemaFormat.parse(rootNode, databasePattern),
+                    null, jsonDynamicSchemaFormat.parse(rootNode, tablePattern), rowSize, dataSize);
```
rowSize and dataSize should be rowSize = 1 and dataSize is byte length of dirtyData

gong commented on inlong-sort/sort-connectors/base/src/main/java/org/apache/inlong/sort/base/metric/sub/SinkTableMetricData.java:
```diff
@@ -50,6 +50,11 @@ public class SinkTableMetricData extends SinkMetricData implements SinkSubMetric
      */
     private final Map<String, SinkMetricData> subSinkMetricMap = Maps.newHashMap();
 
+    /**
+     * The sub sink metric data container of sink metric data
+     */
+    private final Map<String, SinkMetricData> dirtySinkMetricMap = Maps.newHashMap();
+
```
remove it, why add it?

gong commented on inlong-sort/sort-connectors/doris/src/main/java/org/apache/inlong/sort/doris/table/DorisDynamicSchemaOutputFormat.java:
```diff
@@ -507,8 +512,35 @@ private void handleDirtyData(Object dirtyData, DirtyType dirtyType, Exception e)
-            JsonNode rootNode = jsonDynamicSchemaFormat.deserialize(((RowData) dirtyData).getBinary(0));
             metricData.outputDirtyMetricsWithEstimate(
                     jsonDynamicSchemaFormat.parse(rootNode, databasePattern),
                     null, jsonDynamicSchemaFormat.parse(rootNode, tablePattern), 1,
```
schema is null, It will always lost table level metric

yunqingmoswu commented on inlong-sort/sort-connectors/doris/src/main/java/org/apache/inlong/sort/doris/table/DorisDynamicSchemaOutputFormat.java:
```diff
@@ -503,6 +512,42 @@ private void handleDirtyData(Object dirtyData, DirtyType dirtyType, Exception e)
+        try {
+            rootNode = jsonDynamicSchemaFormat.deserialize(((RowData) dirtyData).getBinary(0));
+        } catch (Exception ex) {
+            LOG.warn("handle dirty data: rootnode is null", ex);
```
It is needed to sink dirty data
